### PR TITLE
fri layer

### PR DESCRIPTION
**Pull request by zk-steve**
# Check List

- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] Check every test passed.
- [x] Did you split imports into std, third parties, and custom parts?
- [x] Format your code.
  - [x] cargo fmt
  - [x] cargo clippy
  - [x] `taplo fmt --config taplo/taplo.toml`
- [x] Check unwrap, expect().
- [x] Check clone().
- [x] Check your commit messages.
- [x] Have you added meaningful comments.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -4,17 +4,24 @@ version = "0.1.0"
...
 async-trait = { version = "0.1.81" }
 dotenv = { version = "0.15.0" }
 color-eyre = { version = "0.6.1" }
 thiserror = { version = "1.0.63" }
 rand_core = { version = "0.5.1" }
+regex = { version = "1.10.6" }
+serde_json = { version = "1.0.120" }
+serde = { version = "1.0.204", features = ["derive"] }
+lazy_static = "1.4.0"
+[dev-dependencies]
```
Comment (zk-steve) [inline Cargo.toml:23]: new empty line

File: src/bin/main.rs [hunk 2]
```diff
@@ -1,60 +1,184 @@
...
+use verifier_onchain_services::contracts_caller::verify_fri::verify_merkle::verify_merkle;
+use verifier_onchain_services::contracts_caller::verify_merkle::merkle_statement::verify_merkle_statement;
+use verifier_onchain_services::contracts_caller::verify_merkle::register_fact_merkle::register_fact_merkle;
+use verifier_onchain_services::contracts_caller::verify_merkle::types::{MerkleVerifyInput, RegisterFactVerifyMerkle, VerifyMerkleTransactionInput};
 
 #[tokio::main]
-async fn main() {
-    let (verify_input, root_hash) = sample1();
+async fn main() -> anyhow::Result<()> {
+
```
Comment (zk-steve) [inline src/bin/main.rs]: remove this line

File: src/bin/main.rs [hunk 3]
```diff
@@ -1,60 +1,184 @@
...
-            break;
+    for i in 1..4 {
+        let file_path = format!("./src/data_samples/merkle_verify/merkle_verify_{}.json", i);
+        let input_file = File::open(file_path)?;
+        let reader = BufReader::new(input_file);
+        let merkle_verify_input: MerkleVerifyInput = serde_json::from_reader(reader)?;
+        let verify_merkle_input = || {
+            let mut merkle_view_vec = vec![];
+            for i in 0..merkle_verify_input.merkle_view.len() {
+                merkle_view_vec.push(MoveValue::U256(U256::from_str(&*merkle_verify_input.merkle_view[i].clone()).unwrap()));
```
Comment (zk-steve) [inline src/bin/main.rs]:
```suggestion
                merkle_view_vec.push(MoveValue::U256(U256::from_str(&merkle_verify_input.merkle_view[I]).unwrap()));
```

File: src/bin/main.rs [hunk 4]
```diff
@@ -1,60 +1,184 @@
...
-    loop {
-        n_queries = compute_next_layer(&config, &input_compute).await.expect("E");
-        if !compute_next_layer_view(&config).await.unwrap() {
-            break;
+    for i in 1..4 {
+        let file_path = format!("./src/data_samples/merkle_verify/merkle_verify_{}.json", i);
+        let input_file = File::open(file_path)?;
+        let reader = BufReader::new(input_file);
+        let merkle_verify_input: MerkleVerifyInput = serde_json::from_reader(reader)?;
+        let verify_merkle_input = || {
```
Comment (zk-steve) [inline src/bin/main.rs]: what is the purpose of this function ?

File: src/contracts_caller/types.rs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    }
+}
+
+
+
+
+
+
+
+
```
Comment (zk-steve) [inline src/contracts_caller/types.rs]:
```suggestion
```

File: src/contracts_caller/transaction_helper.rs [hunk 6]
```diff
@@ -32,14 +26,14 @@ pub fn build_transaction(payload: TransactionPayload, sender: &LocalAccount, cha
     )
         .sender(sender.address())
         .sequence_number(i)
-        .max_gas_amount(10000)
+        .max_gas_amount(100000)
```
Comment (zk-steve) [inline src/contracts_caller/transaction_helper.rs]:
```suggestion
        .max_gas_amount(30000)
```

File: src/contracts_caller/verify_fri/compute_next_layer.rs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                    MoveValue::U256(data.evaluation_point.clone()),
+                    MoveValue::U256(data.fri_coset_size.clone()),
+                ]
+            ),
+        ));
+    let tx = build_simulated_transaction(payload, &config.account, config.chain_id);
+    let simulate = config.client.simulate(&tx).await?.into_inner();
+    let vm_status = simulate.get(0).unwrap().info.vm_status.as_str();
+    let vm_status: VmStatus = vm_status.try_into()?;
+    if vm_status.reason == "ECOMPUTE_NEXT_LAYER_NOT_INITIATED" {
```
Comment (zk-steve) [inline src/contracts_caller/verify_fri/compute_next_layer.rs]: const ECOMPUTE_NEXT_LAYER_NOT_INITIATED

File: src/contracts_caller/verify_fri/compute_next_layer.rs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        println!("compute_next_layer check passes");
```
Comment (zk-steve) [inline src/contracts_caller/verify_fri/compute_next_layer.rs]:
```suggestion
```

File: src/contracts_caller/verify_fri/register_fact_fri.rs [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+                &vec![
+                    MoveValue::U256(data.data_to_hash),
+                    MoveValue::U256(data.fri_queue_ptr),
+                    MoveValue::U256(n_queries),
+                ]
+            ),
+        ));
+    let tx = build_transaction(payload, &config.account, config.chain_id);
+    let transaction = config.client.submit_and_wait(&tx).await?.into_inner();
+    let vm_status = transaction.success();
```
Comment (zk-steve) [inline src/contracts_caller/verify_fri/register_fact_fri.rs]: different name for vm_status

File: src/contracts_caller/verify_fri/verify_fri.rs [hunk 10]
```diff
@@ -1,40 +1,49 @@
...
             Identifier::new("verify_fri").unwrap(),
             vec![],
             serialize_values(
                 &vec![
-                    data.proof,
-                    data.fri_queue,
-                    data.evaluation_point,
-                    data.fri_step_size,
-                    data.expected_root,
+                    data.proof.clone(),
```
Comment (zk-steve) [inline src/contracts_caller/verify_fri/verify_fri.rs]: clone ?

File: src/contracts_caller/verify_fri/verify_fri.rs [hunk 11]
```diff
@@ -1,40 +1,49 @@
...
+                    data.expected_root.clone(),
                 ]
             ),
         ));
     let tx = build_transaction(payload, &config.account, config.chain_id);
     let transaction = config.client.submit_and_wait(&tx).await?.into_inner();
 
-    let event_type = MoveType::from_str(&format!("{}::fri_statement::FriCtx", config.module_address)).unwrap();
+    let event_type = MoveType::from_str(&format!("{}::fri_statement_contract::FriCtx", config.module_address)).unwrap();
     let fri_ctx_data = get_event_from_transaction(transaction.clone(), event_type).await?;
```
Comment (zk-steve) [inline src/contracts_caller/verify_fri/verify_fri.rs]: no need to await for `get_event_from_transaction`

File: src/error.rs [hunk 12]
```diff
@@ -1,11 +1,28 @@
...
+
+    #[error("transaction not succeed {0}")]
+    TransactionNotSucceed(String),
 
     #[error("not found")]
     NotFound,
+
+    #[error("property not found")]
+    PropertyNotFound,
+
```
Comment (zk-steve) [inline src/error.rs]:
```suggestion
```

File: src/verify_flow.rs [hunk 13]
```diff
@@ -1,63 +1,208 @@
...
+    use crate::contracts_caller::verify_fri::register_fact_fri::register_fact_fri;
+    use crate::contracts_caller::verify_fri::types::{ComputeNextLayer, FriVerifyInput, InitFriGroup, RegisterFactVerifyFri, VerifyFriTransactionInput};
+    use crate::contracts_caller::verify_fri::verify_fri::verify_fri;
+    use crate::contracts_caller::verify_fri::verify_merkle::verify_merkle;
+    use crate::contracts_caller::verify_merkle::merkle_statement::verify_merkle_statement;
+    use crate::contracts_caller::verify_merkle::register_fact_merkle::register_fact_merkle;
+    use crate::contracts_caller::verify_merkle::types::{MerkleVerifyInput, RegisterFactVerifyMerkle, VerifyMerkleTransactionInput};
 
     #[tokio::test]
     pub async fn test_a() -> anyhow::Result<()> {
```
Comment (zk-steve) [inline src/verify_flow.rs]: rename

File: src/contracts_caller/verify_merkle/types.rs
Comment (zk-steve) [inline src/contracts_caller/verify_merkle/types.rs:1]: split into multiple files with the file name = struct name

File: src/verify_flow.rs [hunk 14]
```diff
@@ -1,63 +1,208 @@
...
+        let node_url = aptos_container.get_node_url().await.unwrap();
+        let sender_account = LocalAccount::from_private_key("0xa7599766d8aaace6959eb7e315c1c76af44276641dff8912c9356e3d0799c80d", 0).unwrap();
+        let module_account = LocalAccount::from_private_key("0x73791ce34b2414d4afcb87561b0c442e48a3260f1c96de31da80f7cf2eec8113", 0).unwrap();
+
+        aptos_container.faucet(&module_account).await.unwrap();
+        aptos_container.faucet(&sender_account).await.unwrap();
+
+        let config = AppConfig::from(EnvConfig {
+            node_url,
+            private_key: "0xa7599766d8aaace6959eb7e315c1c76af44276641dff8912c9356e3d0799c80d".to_string(),
```
Comment (zk-steve) [inline src/verify_flow.rs]: get private key from `sender_account`

File: src/verify_flow.rs
Comment (zk-steve) [inline src/verify_flow.rs:1]: rename to `flow_test.rs`

File: src/verify_flow.rs [hunk 15]
```diff
@@ -1,63 +1,208 @@
...
+            module_address: module_account.address().to_string(),
+            chain_id: "4".to_string(),
+        });
+
         let sequence_number = config.client.get_account(config.account.address()).await?.into_inner().sequence_number;
         config.account.set_sequence_number(sequence_number);
 
+        let mut named_addresses = HashMap::new();
+        named_addresses.insert("verifier_addr".to_string(), module_account.address().to_string());
+        aptos_container.upload_contract("./contract-sample/navori", &module_account, &named_addresses, true).await.unwrap();
```
Comment (zk-steve) [inline src/verify_flow.rs]: fix contract path

File: src/bin/main.rs [hunk 16]
```diff
@@ -1,60 +1,184 @@
...
+        };
+        let (merkle_view, initial_merkle_queue, height, expected_root) = verify_merkle_input();
+        let verify_merkle_input = VerifyMerkleTransactionInput {
+            merkle_view,
+            initial_merkle_queue,
+            height,
+            expected_root,
+        };
+
+        let sequence_number = config.client.get_account(config.account.address()).await?.into_inner().sequence_number;
```
Comment (zk-steve) [inline src/bin/main.rs:58]: only need to do this once

File: src/bin/main.rs [hunk 17]
```diff
@@ -1,60 +1,184 @@
...
+        config.account.set_sequence_number(sequence_number);
```
Comment (zk-steve) [inline src/bin/main.rs]: only need to do this once

File: src/verify_flow.rs [hunk 18]
```diff
@@ -1,63 +1,208 @@
...
+        let module_account = LocalAccount::from_private_key("0x73791ce34b2414d4afcb87561b0c442e48a3260f1c96de31da80f7cf2eec8113", 0).unwrap();
+
+        aptos_container.faucet(&module_account).await.unwrap();
+        aptos_container.faucet(&sender_account).await.unwrap();
+
+        let config = AppConfig::from(EnvConfig {
+            node_url,
+            private_key: "0xa7599766d8aaace6959eb7e315c1c76af44276641dff8912c9356e3d0799c80d".to_string(),
+            module_address: module_account.address().to_string(),
+            chain_id: "4".to_string(),
```
Comment (zk-steve) [inline src/verify_flow.rs]: chain_id: TESTING.id().to_string(),

File: src/contracts_caller/transaction_helper.rs [hunk 19]
```diff
@@ -3,24 +3,18 @@ use std::time::SystemTime;
...
-pub fn str_to_u256(s: &str) -> MoveValue {
-    let u256_value = U256::from_str(s).unwrap();
-    MoveValue::U256(u256_value)
-}
+use crate::error::CoreError;
 
-pub fn str_to_bool(s: &str) -> bool {
-    let bool_str = s.trim_start_matches("Bool(").trim_end_matches(")");
-    bool::from_str(bool_str).unwrap()
+pub fn str_to_u256(s: &str) -> Result<U256, CoreError> {
```
Comment (zk-steve) [inline src/contracts_caller/transaction_helper.rs:17]: add 
#[inline]

File: src/flow_test.rs [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+                return Ok(());
+            }
+
+            if !register_fact_merkle(&config, input_register_fact_merkle).await? {
+                eprintln!("something went wrong!");
+                return Ok(());
+            }
+        }
+
+        for i in 1..8 {
```
Comment (zk-steve) [inline src/flow_test.rs]: split this into a separate function

Comment (zk-steve) [inline src/flow_test.rs]: parallelize this

File: src/contracts_caller/verify_merkle/merkle_statement.rs [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+use aptos_sdk::move_types::language_storage::ModuleId;
+use aptos_sdk::move_types::value::serialize_values;
+use aptos_sdk::rest_client::aptos_api_types::{Event, MoveType};
+use aptos_sdk::types::transaction::{EntryFunction, TransactionPayload};
+
+use crate::config::AppConfig;
+use crate::contracts_caller::transaction_helper::{build_transaction, get_event_from_transaction};
+use crate::contracts_caller::verify_merkle::types::verify_merkle_input::VerifyMerkleTransactionInput;
+
+pub async fn verify_merkle_statement(config: &AppConfig, data: VerifyMerkleTransactionInput) -> anyhow::Result<(Event, Event)> {
```
Comment (zk-steve) [inline src/contracts_caller/verify_merkle/merkle_statement.rs]: return (VerifyMerkle, RegisterFactVerifyMerkle) instead

File: src/contracts_caller/verify_merkle/types/register_fact_verify_merkle.rs [hunk 22]
```diff
@@ -0,0 +1,26 @@
+use aptos_sdk::move_types::u256::U256;
+use aptos_sdk::rest_client::aptos_api_types::Event;
+use crate::contracts_caller::transaction_helper::str_to_u256;
+use crate::error::CoreError;
+use crate::error::CoreError::PropertyNotFound;
+
+#[derive(Debug)]
+pub struct RegisterFactVerifyMerkle{
+    pub channel_ptr: U256,
```
Comment (zk-steve) [inline src/contracts_caller/verify_merkle/types/register_fact_verify_merkle.rs]: always = 339

Comment (zk-steve) [reply src/contracts_caller/verify_merkle/types/register_fact_verify_merkle.rs]: remove it

File: .gitmodules [hunk 23]
```diff
@@ -0,0 +1,6 @@
+[submodule "navori"]
+	path = navori
+	url = git@github.com:sota-zk-labs/navori.git
+[submodule "contract-sample/navori"]
```
Comment (zk-steve) [inline .gitmodules:1]: remove this

File: contract-sample/navori
Comment (zk-steve) [inline contract-sample/navori:1]: remove

File: src/contracts_caller/verify_fri/fmul.rs
Comment (zk-steve) [inline src/contracts_caller/verify_fri/fmul.rs:1]: re move

File: src/bin/main.rs [hunk 24]
```diff
@@ -1,60 +1,49 @@
...
+            sample_verify_merkle_input(i)?;
+        verify_merkle(
+            &config,
+            merkle_view,
+            initial_merkle_queue,
+            height,
+            expected_root,
+        )
+        .await?;
+        eprintln!("Verify Merkle {} success", i);
```
Comment (zk-steve) [inline src/bin/main.rs]: log.Log

File: src/bin/main.rs [hunk 25]
```diff
@@ -1,60 +1,49 @@
...
+            &config,
+            fri_verify_input,
+            proof,
+            fri_queue,
+            evaluation_point,
+            fri_step_size,
+            expected_root,
+        )
+        .await?;
+        eprintln!("Verify FRI {} success", i);
```
Comment (zk-steve) [inline src/bin/main.rs]: find all eprintln and fix

File: src/contracts_caller/verify_merkle/types/register_fact_verify_merkle.rs [hunk 26]
```diff
@@ -0,0 +1,50 @@
+use crate::contracts_caller::transaction_helper::{str_to_u256, str_to_u64};
```
Comment (zk-steve) [inline src/contracts_caller/verify_merkle/types/register_fact_verify_merkle.rs:4]: format imports

File: src/flow_test.rs [hunk 27]
```diff
@@ -0,0 +1,103 @@
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use aptos_sdk::crypto::ValidCryptoMaterialStringExt;
+    use aptos_sdk::types::chain_id::NamedChain::TESTING;
+    use aptos_sdk::types::LocalAccount;
+    use aptos_testcontainer::test_utils::aptos_container_test_utils::{lazy_aptos_container, run};
+    use log::info;
+    use crate::config::AppConfig;
```
Comment (zk-steve) [inline src/flow_test.rs:11]: new line

**Review by Draply — APPROVED** [review]

**Pull request merged**